### PR TITLE
Cache decisions for ColumnVisibility filters

### PR DESCRIPTION
**Pull request by apmoriarty**
- The TokenFilterBase now has a column visibility cache to reduce the time spent parsing visibilities. 
- The cache has a default size of 50 but the size can be overridden using the COLUMN_VISIBILITY_CACHE_SIZE age off parameter. 
- added unit tests for the ColumnVisibilityOrFilter and ColumnVisibilityAndFilter

Addresses https://github.com/NationalSecurityAgency/datawave/issues/1579

### CODE REVIEW

**Comment by avgAGB** [comment]
You probably already have something like it in the works, but would it also be possible to test the functionality of the ColumnVisibilityAndFilter? It may not even need it's own test class per se, an instance of it in the existing unit test with a few examples tested might do the trick.

File: warehouse/core/src/main/java/datawave/iterators/filter/ColumnVisibilityAndFilter.java [hunk 1]
```diff
@@ -6,19 +6,21 @@
...
+     *
      * @param k
      *            {@code Key} object containing the row, column family, and column qualifier.
      * @param v
      *            {@code Value} object containing the value corresponding to the {@code Key: k}
-     * @return {@code String} object containing the column visibility of {@code k}
+     * @return true if the provided key's column visibility matches the configured {@link TokenFilterBase}'s pattern
      */
     @Override
     public boolean hasToken(Key k, Value v, byte[][] testTokens) {
```
Comment (FineAndDandy) [inline warehouse/core/src/main/java/datawave/iterators/filter/ColumnVisibilityAndFilter.java]: most of this method can be moved to an abstract method in TokenFilterBase

**Review by FineAndDandy — CHANGES_REQUESTED** [review]

File: warehouse/core/src/test/java/datawave/iterators/filter/ColumnVisibilityOrFilterTest.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Test for the {@link ColumnVisibilityOrFilter}
+ */
+public class ColumnVisibilityOrFilterTest {
```
Comment (hlgp) [inline warehouse/core/src/test/java/datawave/iterators/filter/ColumnVisibilityOrFilterTest.java]: How about an AndFilterTest, too?

File: warehouse/core/src/main/java/datawave/iterators/filter/ColumnVisibilityAndFilter.java (hunk 1 above)
Comment (apmoriarty) [reply warehouse/core/src/main/java/datawave/iterators/filter/ColumnVisibilityAndFilter.java]: done

File: warehouse/core/src/test/java/datawave/iterators/filter/ColumnVisibilityOrFilterTest.java (hunk 2 above)
Comment (apmoriarty) [reply warehouse/core/src/test/java/datawave/iterators/filter/ColumnVisibilityOrFilterTest.java]: done

**Review by ivakegg — DISMISSED** [review]

File: warehouse/age-off/src/main/java/datawave/iterators/filter/ColumnVisibilityOrFilter.java [hunk 3]
```diff
@@ -12,25 +12,23 @@
 public class ColumnVisibilityOrFilter extends TokenFilterBase {
```
Comment (matthpeterson) [inline warehouse/age-off/src/main/java/datawave/iterators/filter/ColumnVisibilityOrFilter.java:10]: We may want to deprecate this class.  I think we tested https://github.com/NationalSecurityAgency/datawave/blob/integration/warehouse/age-off/src/main/java/datawave/iterators/filter/ColumnVisibilityTokenizingFilter.java and found it to be much faster.  Perhaps this caching change could be made to it.

Comment (apmoriarty) [reply warehouse/age-off/src/main/java/datawave/iterators/filter/ColumnVisibilityOrFilter.java:10]: I can deprecate the ColumnVisibilityOrFilter and ColumnVisibilityAndFilter.

~~It looks like the ColumnVisibilityTokenizingFilter ultimately delegates to the `accept()` method found in the TokenFilterBase.~~

The TokenizingFilterBase could also use a cache. I can add that if you like.

**Review by avgAGB — APPROVED** [review]

**Comment by apmoriarty** [comment]
Local testing showed some improvements in some cases, on-cluster testing showed degradation in the common cases.

**Pull request closed without merging**